command/windows_service_install.go

```diff
@@ -0,0 +1,347 @@
+
+	// Now let's install that service
+	if err := c.serviceInstall(m, opts); err != nil {
+		return fmt.Errorf("service install failed - %w", err)
+	}
+
+	return nil
+}
+
+func (c *WindowsServiceInstallCommand) serviceInstall(m winsvc.WindowsServiceManager, opts *windowsInstallOpts) error {
+	var err error
+	var srvc winsvc.WindowsService
+	cmd := fmt.Sprintf("%s agent -config %s", opts.binaryPath, opts.configDir)
+
+	exists, err := m.IsServiceRegistered(winsvc.WINDOWS_SERVICE_NAME)
+	if err != nil {
+		return fmt.Errorf("service registration check failed - %w", err)
+	}
+
+	// If the service already exists, open it and update. Otherwise
+	// create a new service.
+	if exists {
+		srvc, err = m.GetService(winsvc.WINDOWS_SERVICE_NAME)
+		if err != nil {
+			return fmt.Errorf("unable to get existing service - %w", err)
+		}
+		defer srvc.Close()
+		if err := srvc.Configure(winsvc.WindowsServiceConfiguration{
+			StartType:      winsvc.StartAutomatic,
+			DisplayName:    winsvc.WINDOWS_SERVICE_DISPLAY_NAME,
+			Description:    winsvc.WINDOWS_SERVICE_DESCRIPTION,
+			BinaryPathName: cmd,
+		}); err != nil {
+			return fmt.Errorf("unable to configure service - %w", err)
+		}
+	} else {
+		srvc, err = m.CreateService(winsvc.WINDOWS_SERVICE_NAME, opts.binaryPath,
+			winsvc.WindowsServiceConfiguration{
+				StartType:      winsvc.StartAutomatic,
+				DisplayName:    winsvc.WINDOWS_SERVICE_DISPLAY_NAME,
+				Description:    winsvc.WINDOWS_SERVICE_DESCRIPTION,
+				BinaryPathName: cmd,
+			},
+		)
+		if err != nil {
+			return fmt.Errorf("unable to create service - %w", err)
+		}
+		defer srvc.Close()
+	}
+
+	// Enable the service in the Windows eventlog
+	if err := srvc.EnableEventlog(); err != nil {
+		return fmt.Errorf("could not configure eventlog - %w", err)
+	}
+
+	// Ensure the service is stopped
+	if err := srvc.Stop(); err != nil {
+		return fmt.Errorf("could not stop service - %w", err)
+	}
+
+	// Start the service so the new binary is in use
+	if err := srvc.Start(); err != nil {
+		return fmt.Errorf("could not start service - %w", err)
+	}
+
+	return nil
+}
+
+func (c *WindowsServiceInstallCommand) configInstall(opts *windowsInstallOpts) error {
+	// If the config or data directory are unset, default them
+	if opts.configDir == "" {
+		opts.configDir = filepath.Join(winsvc.WINDOWS_INSTALL_APPDATA_DIRECTORY, "config")
+	}
+	if opts.dataDir == "" {
+		opts.dataDir = filepath.Join(winsvc.WINDOWS_INSTALL_APPDATA_DIRECTORY, "data")
+	}
+
+	var err error
+	if opts.configDir, err = c.expandPathFn(opts.configDir); err != nil {
+		return fmt.Errorf("cannot generate configuration path - %s", err)
+	}
+	if opts.dataDir, err = c.expandPathFn(opts.dataDir); err != nil {
+		return fmt.Errorf("cannot generate data path - %s", err)
+	}
+
+	// Ensure directories exist
+	if err = os.MkdirAll(opts.configDir, 0o755); err != nil {
+		return fmt.Errorf("cannot create configuration directory - %s", err)
+	}
```

Comment:
I'm not sure what these permissions numbers strings even mean in Windows, but the config dir will often have certs in it so we should make sure that's not world-readable. (The datadir is a bit more "it's complicated" because of the alloc directories which need to be readable by workloads.)

Reply:
The mode is meaningless for Windows and is ignored. With regards to the actual permissions applied to the directories, I've updated how the directories are created so that if the command is creating the directory during the install process (config and data directories) it will apply an ACL that restricts access to the current user and the Administrators group.